### PR TITLE
Split survey results per facilitator

### PR DESCRIPTION
**Pull request by hacodeorg**
This change in the new survey pipeline enables displaying facilitator results per facilitator. It also enables permission check, i.e. facilitator cannot see other facilitator's result while program manager and workshop admin can see all results.

This change currently impacts only viewing CSF 201 Deep Dive workshop survey results.

## Examples
**Before** (from [real data](https://studio.code.org/pd/workshop_dashboard/daily_survey_results/6482) in production): All facilitator responds are combined. 
Results can be viewed by `workshop_admin`, `program_manager` and `facilitator`.

Screen Shot 2019-06-05 at 5 13 36 PM

**After** (development env): Results are displayed per-facilitator. 
`workshop_admin` and `program_manager` can see all facilitator results.
Screen Shot 2019-06-06 at 3 59 41 PM

`facilitator` can only see his/her own results
Screen Shot 2019-06-06 at 3 57 32 PM

## How tested
- Unit tests: Run all tests for new survey pipeline
  - `bundle exec rails test test/lib/pd/survey_pipeline/daily_survey_retriever_test.rb test/lib/pd/survey_pipeline/daily_survey_parser_test.rb test/lib/pd/survey_pipeline/daily_survey_joiner_test.rb test/lib/pd/survey_pipeline/mapper_test.rb test/lib/pd/survey_pipeline/reducer_test.rb test/lib/pd/survey_pipeline/daily_survey_decorator_test.rb test/controllers/api/v1/pd/workshop_survey_report_controller_test.rb`

- Manual tests: 
  - Download survey responds for pre/post/facilitator CSF201 forms for workshop 6482 to Pd::SurveyQuestion and Pd::WorkshopDailySurvey and Pd::WorkshopFacilitatorDailySurvey.
  - API check: `http://localhost-studio.code.org:3000/api/v1/pd/workshops/6482/generic_survey_report`
  - UI check: `http://localhost-studio.code.org:3000/pd/workshop_dashboard/daily_survey_results/6482`

### CODE REVIEW

File: dashboard/lib/pd/survey_pipeline/daily_survey_decorator.rb [hunk 1]
```diff
@@ -37,45 +37,83 @@ def self.decorate(summary_data:, parsed_data:)
...
 
-        result[:this_workshop][form_name][:general][summary[:name]] = summary[:reducer_result]
+        # Create an entry in result[:this_workshop].
+        # General format:
+        # Hash{form_name => {response_count => number, general => {question_name => summary_result}}}
+        #
+        # Format for facilicator-specific result:
+        # Hash{form_name => {response_count => number,
+        #   facilitator => {question_name => {factilitator_name => summary_result}}}}
+        if summary[:facilitator_id]
```
Comment (clareconstantine) [inline dashboard/lib/pd/survey_pipeline/daily_survey_decorator.rb:72]: does this take into account permissions, that co-facilitators should not see each other's results?

Comment (agealy) [reply dashboard/lib/pd/survey_pipeline/daily_survey_decorator.rb:72]: +1

Comment (hacodeorg) [reply dashboard/lib/pd/survey_pipeline/daily_survey_decorator.rb:72]: oh great point, it does not. Will add

Comment (hacodeorg) [reply dashboard/lib/pd/survey_pipeline/daily_survey_decorator.rb:72]: Added permission filter and some unit tests. I may need help to test an end-to-end scenario locally.

**Review by clareconstantine — APPROVED** [review]

**Comment by clareconstantine** [comment]
could you specify which workshops/surveys this impacts, so we remember looking back at this PR later?

**Review by agealy — APPROVED** [review]

**Comment by hacodeorg** [comment]
This currently impacts only CSF 201 (Deep dive) workshop.
(Will also add this to the PR description.)

> could you specify which workshops/surveys this impacts, so we remember looking back at this PR later?

**Pull request merged**